<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="xcmsTutorials">
<title>Exploring and Analyzing LC-MS data with *Spectra* and *xcms* • xcmsTutorials</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Exploring and Analyzing LC-MS data with *Spectra* and *xcms*">
<meta property="og:description" content="xcmsTutorials">
<meta property="og:image" content="https://jorainer.github.io/xcmsTutorials/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">xcmsTutorials</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/xcms-preprocessing.html">Exploring and Analyzing LC-MS data with *Spectra* and *xcms*</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jorainer/xcmsTutorials/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Exploring and Analyzing LC-MS data with *Spectra* and *xcms*</h1>
                        <h4 data-toc-skip class="author">Johannes
Rainer</h4>
            <address class="author_afil">
      Eurac Research, Bolzano, Italy; <a href="mailto:johannes.rainer@eurac.edu" class="email">johannes.rainer@eurac.edu</a> github: jorainer twitter:
jo_rainer<br><h4 data-toc-skip class="date">May 17 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jorainer/xcmsTutorials/blob/HEAD/vignettes/xcms-preprocessing.Rmd" class="external-link"><code>vignettes/xcms-preprocessing.Rmd</code></a></small>
      <div class="d-none name"><code>xcms-preprocessing.Rmd</code></div>
    </address>
</div>

    
    
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p>In this document we discuss mass spectrometry (MS) data handling and
exploration using the <em><a href="https://bioconductor.org/packages/3.18/MsExperiment" class="external-link">MsExperiment</a></em>
and <em><a href="https://bioconductor.org/packages/3.18/Spectra" class="external-link">Spectra</a></em>
Bioconductor packages and perform the preprocessing of a small liquid
chromatography (LC)-MS data set <em><a href="https://bioconductor.org/packages/3.18/xcms" class="external-link">xcms</a></em>
package. In addition we use functionality from the <em><a href="https://bioconductor.org/packages/3.18/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></em>
and <em><a href="https://bioconductor.org/packages/3.18/MsCoreUtils" class="external-link">MsCoreUtils</a></em>
package for general tasks frequently performed in metabolomics data
analysis. The preprocessing of the LC-MS data comprises chromatographic
peak detection, sample alignment and peak correspondence. Particular
emphasis is given on deriving and defining data-set dependent values for
the most crucial parameters of popular preprocessing methods.</p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Preprocessing is the first step in the analysis of
<em>untargeted</em> LC-MS or gas chromatography (GC)-MS data. The aim of
the preprocessing is the quantification of signals from ions measured in
a sample, adjusting for any potential LC drifts between samples and the
matching of the quantified signal across samples within an experiment.
The resulting two-dimensional matrix with abundances of the so called
<em>LC-MS features</em> in all samples can then be further processed,
e.g. by normalizing the data to remove differences due to sample
processing, batch effects or injection order-dependent signal drifts.
After preprocessing, the LC-MS features usually only characterized by
their mass-to-charge ration (m/z) and retention time, have to be
annotated to the actual ions and metabolites they represent. Data
normalization and annotation are not covered in this document.</p>
<div class="section level3">
<h3 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h3>
<p>The analysis in this document requires an R version &gt;= 4.3.0 and
recent versions of the <em><a href="https://bioconductor.org/packages/3.18/MsExperiment" class="external-link">MsExperiment</a></em>
and <em><a href="https://bioconductor.org/packages/3.18/Spectra" class="external-link">Spectra</a></em>
Bioconductor packages as well as the developmental version of the <em><a href="https://bioconductor.org/packages/3.18/xcms" class="external-link">xcms</a></em>
package. These can be installed using the code below.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Install the Bioconductor package manager</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Install the required packages</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"msdata"</span>,</span>
<span>                       <span class="st">"Spectra"</span>,</span>
<span>                       <span class="st">"MsExperiment"</span>,</span>
<span>                       <span class="st">"MetaboCoreUtils"</span>,</span>
<span>                       <span class="st">"MsCoreUtils"</span>,</span>
<span>                       <span class="st">"png"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sneumann/xcms"</span><span class="op">)</span></span></code></pre></div>
<p>The <a href="https://github.com/jorainer/metabolomics2018/blob/master/xcms-preprocessing.Rmd" class="external-link">xcms-preprocessing.Rmd</a>
file with all the code for the analysis can be downloaded from [github]
(<a href="https://github.com/jorainer/metabolomics2018" class="external-link uri">https://github.com/jorainer/metabolomics2018</a>) ideally
with <code>git clone https://github.com/jorainer/metabolomics2018</code>
from the command line.</p>
</div>
<div class="section level3">
<h3 id="mass-spectrometry">Mass spectrometry<a class="anchor" aria-label="anchor" href="#mass-spectrometry"></a>
</h3>
<p>Mass spectrometry allows to measure abundances of charged molecules
(ions) in a sample. Abundances are determined as ion counts for a
specific mass-to-charge ratio m/z. The measured signal is represented as
a spectrum: intensities along m/z.</p>
<p><img src="MS.png"></p>
<p>Many ions will result, when measured with MS alone, in a very similar
m/z making it difficult or impossible to discriminate them. MS is thus
frequently coupled with a second technology to separate them prior
quantification based on properties other than their mass (e.g. based on
their polarity). Common choices are gas chromatography (GC) or liquid
chromatography (LC). In a typical LC-MS setup a samples gets injected
into the system, molecules are separated in the LC column while the MS
instruments continuously (at discrete time points) continues to measure
all ions that get generated from the molecules eluting at different time
points from the LC-column. Molecules get thus separated on two different
dimensions, the retention time dimension (from the LC) and the
mass-to-charge dimension (from the MS) making it easier to measure and
identify molecules in more complex samples.</p>
<p><img src="LCMS.png"></p>
<p>In such GC/LC-MS based untargeted metabolomics experiments the data
is analyzed along the retention time dimension and
<em>chromatographic</em> peaks (which are supposed to represent the
signal from ions of a certain type of molecule) are quantified.</p>
</div>
<div class="section level3">
<h3 id="definitions-and-common-naming-convention">Definitions and common naming convention<a class="anchor" aria-label="anchor" href="#definitions-and-common-naming-convention"></a>
</h3>
<p>Naming conventions and terms used in this document are:</p>
<ul>
<li>
<em>chromatographic peak</em>: peak containing the signal from an
ion in retention time dimension (different from a <em>mass</em> peak
that represents the signal along the m/z dimension within a
spectrum).</li>
<li>
<em>chromatographic peak detection</em>: process in which
chromatographic peaks are identified within each file.</li>
<li>
<em>alignment</em>: process that adjusts for retention time
differences (i.e. possible signal drifts from the LC) between
measurements/files.</li>
<li>
<em>correspondence</em>: grouping of chromatographic peaks
(presumably from the same ion) across files.</li>
<li>
<em>feature</em> (or <em>LC-MS features</em>): entity representing
signal from the same type of ion/molecule, characterized by its specific
retention time and m/z. In <em>xcms</em>, features represent identified
chromatographic peaks grouped across files/samples.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="workflow-exploring-and-analyzing-lc-ms-data-with-spectra-and-xcms">Workflow: exploring and analyzing LC-MS data with <em>Spectra</em>
and <em>xcms</em><a class="anchor" aria-label="anchor" href="#workflow-exploring-and-analyzing-lc-ms-data-with-spectra-and-xcms"></a>
</h2>
<p>This workflow describes the basic data handling (I/O) of mass
spectrometry data using the <em><a href="https://bioconductor.org/packages/3.18/MsExperiment" class="external-link">MsExperiment</a></em>
and <em><a href="https://bioconductor.org/packages/3.18/Spectra" class="external-link">Spectra</a></em>
package, and the LC-MS data preprocessing using <em><a href="https://bioconductor.org/packages/3.18/xcms" class="external-link">xcms</a></em>. The
first part of the workflow is focused on data import, access and
visualization which is followed by the description of a simple data
centroiding approach and finally the <em>xcms</em>-based LC-MS data
preprocessing that comprises chromatographic peak detection, alignment
and correspondence. The workflow does not cover data normalization
procedures, compound identification and differential abundance
analysis.</p>
<div class="section level3">
<h3 id="data-import-and-exploration">Data import and exploration<a class="anchor" aria-label="anchor" href="#data-import-and-exploration"></a>
</h3>
<p>The example data set of this workflow consists of two files in mzML
format with signals from pooled human serum samples measured with a
ultra high performance liquid chromatography (UHPLC) system (Agilent
1290) coupled with a Q-TOF MS (TripleTOF 5600+ AB Sciex) instrument.
Chromatographic separation was based on hydrophilic interaction liquid
chromatography (HILIC) separating metabolites depending on their
polarity. The setup thus allows to measure small polar compounds and
hence metabolites from the main metabolic pathways. The input files
contain all signals measured by the MS instrument (so called <em>profile
mode</em> data). To reduce file sizes, the data set was restricted to an
m/z range from 105 to 134 and retention times from 0 to 260 seconds.</p>
<p>In the code block below we first load all required libraries and
define the location of the mzML files, which are part of the
<em>msdata</em> R package. We also define a <code>data.frame</code>
describing the samples/experiment and pass this to the
<code>readMsExperiment</code> function that imports the data. Similar to
the <em>on-disk-mode</em> described in <span class="citation">(Gatto,
Gibb, and Rainer 2020)</span>, the m/z and intensity values are not
immediately loaded into the memory but only when required which enables
also analyses of very large experiments.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sneumann/xcms" class="external-link">xcms</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MsExperiment" class="external-link">MsExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/Spectra" class="external-link">Spectra</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define the file names.</span></span>
<span><span class="va">fls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"sciex"</span>, package <span class="op">=</span> <span class="st">"msdata"</span><span class="op">)</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define a data.frame with additional information on the files.</span></span>
<span><span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">fls</span><span class="op">)</span>,</span>
<span>                 injection_idx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">19</span><span class="op">)</span>,</span>
<span>                 sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"POOL_1"</span>, <span class="st">"POOL_2"</span><span class="op">)</span>,</span>
<span>                 group <span class="op">=</span> <span class="st">"POOL"</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/readMsExperiment.html" class="external-link">readMsExperiment</a></span><span class="op">(</span><span class="va">fls</span>, sampleData <span class="op">=</span> <span class="va">pd</span><span class="op">)</span></span>
<span><span class="va">data</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Spectra: MS1 (1862) </span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - spectra: 2 sample(s) to 1862 element(s).</span></span></code></pre>
<p>Note that for a <em>real</em> experiment it is suggested to define
such a <em>phenodata</em> table as an e.g. tab delimited text file (or
an xls sheet) that contains all of the raw data file names along with
all the relevant sample information for each file as additional columns.
Such a data table could then be imported into R using
e.g. <code>read.table</code> or <code>read_xlsx</code> (from the
<em>readxl</em> R package) and the file names could be passed to
<code>readMsExperiment</code> with
<code>files = paste0(MZML_PATH, "/", pd$mzML_file)</code> where
<code>MZML_PATH</code> would be a variable specifying the directory
containing the data (mzML) files, <code>pd</code> would be the imported
phenodata table with a column named <code>"mzML_file"</code> containing
the names of the individual raw files.</p>
<p>Next we set up parallel processing. This ensures that all required
cores are registered and available from the beginning of the analysis.
All data access and analysis functions of <em>xcms</em> on a per-file
basis and will use this setup by default.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Set up parallel processing using 2 cores</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">==</span> <span class="st">"unix"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">bpstart</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">bpstart</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The MS data of the experiment is now <em>represented</em> by a
<code>MsExperiment</code> object. Phenotype information can be retrieved
with the <code>sampleData</code> function from that object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Access phenotype information</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2 rows and 5 columns</span></span>
<span><span class="co">##            file injection_idx      sample       group spectraOrigin</span></span>
<span><span class="co">##     &lt;character&gt;     &lt;numeric&gt; &lt;character&gt; &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">## 1 20171016_P...             1      POOL_1        POOL /usr/local...</span></span>
<span><span class="co">## 2 20171016_P...            19      POOL_2        POOL /usr/local...</span></span></code></pre>
<p>The MS data is stored as a <code>Spectra</code> object within the
<code>MsExperiment</code> and can be accessed using the
<code>spectra</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 1862 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##        msLevel     rtime scanIndex</span></span>
<span><span class="co">##      &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1            1     0.280         1</span></span>
<span><span class="co">## 2            1     0.559         2</span></span>
<span><span class="co">## 3            1     0.838         3</span></span>
<span><span class="co">## 4            1     1.117         4</span></span>
<span><span class="co">## 5            1     1.396         5</span></span>
<span><span class="co">## ...        ...       ...       ...</span></span>
<span><span class="co">## 1858         1   258.636       927</span></span>
<span><span class="co">## 1859         1   258.915       928</span></span>
<span><span class="co">## 1860         1   259.194       929</span></span>
<span><span class="co">## 1861         1   259.473       930</span></span>
<span><span class="co">## 1862         1   259.752       931</span></span>
<span><span class="co">##  ... 33 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## 20171016_POOL_POS_1_105-134.mzML</span></span>
<span><span class="co">## 20171016_POOL_POS_3_105-134.mzML</span></span></code></pre>
<p>This <code>Spectra</code> object represents the full LC-MS data of
the experiment. Each element in this object is a spectrum (in one
sample/file) with all information provided by the respective original
data (mzML) file. Spectra are organized linearly, i.e. all spectra from
all files are within the same <code>Spectra</code> object, one after the
other.</p>
</div>
<div class="section level3">
<h3 id="basic-data-access-and-visualization">Basic data access and visualization<a class="anchor" aria-label="anchor" href="#basic-data-access-and-visualization"></a>
</h3>
<p>The MS data of an experiment is stored as a <code>Spectra</code>
object within the <code>MsExperiment</code> and this
<code>MsExperiment</code> contains the experiment’s sample information
and manages the <em>linkage</em> between samples and spectra. The
<code>length</code> of an <code>MsExperiment</code> is defined by the
number of samples within the object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2</span></span></code></pre>
<p>As detailed above, the <code>spectra</code> function can be used to
access the MS data of the experiment. As an example we below calculate
the retention time range of the full experiment. Retention times can be
extracted with the <code>rtime</code> function from the
<code>Spectra</code> object within <code>data</code>. To avoid nested
function calls and hence improve the readability of the code, we use the
R pipe operator <code>|&gt;</code> that allows to concatenate
consecutive calls in a more readable fashion.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]   0.275 259.757</span></span></code></pre>
<p>The <code>Spectra</code> object returned by <code>spectra</code>,
contains all spectra from all files (samples) of an experiment. To
access MS data from only a single file (sample) we need to subset the
<code>MsExperiment</code> object to that particular sample before
extracting the spectra data with <code>spectra</code>. Below we subset
the data to the second sample using the <code>[</code> function, access
the spectra from that sample, extract their retention times and
calculate their range.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]   0.275 259.752</span></span></code></pre>
<p>We can also access a single spectrum from this second sample and
visualize its data. We thus below extract the spectrum number 123 from
the second sample and use the <code>plotSpectra</code> function to plot
that spectrum.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">123</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectra</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></span></code></pre></div>
<p><img src="xcms-preprocessing_files/figure-html/unnamed-chunk-3-1.png" width="672"></p>
<p>We can see several relatively large signals (peaks) in that spectrum
and also a number of low intensity peaks. This spectrum represents a
single scan of the LC-MS data run of the second sample. It’s retention
time can be extracted with <code>rtime</code>, intensity and m/z values
can be extracted with the <code>intensity</code> and <code>mz</code>
functions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 34.314</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">intensity</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NumericList of length 1</span></span>
<span><span class="co">## [[1]] 0 282 0 141 0 0 141 0 141 0 141 0 ... 563 563 422 0 0 282 282 0 282 141 0</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">mz</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NumericList of length 1</span></span>
<span><span class="co">## [[1]] 105.95354942709 105.955001209814 ... 133.105299625013 133.106926815539</span></span></code></pre>
<p>Note that, even for a <code>Spectra</code> with data from a single
spectrum, the <code>intensity</code> and <code>mz</code> functions
return a <code>NumericList</code> with the intensity and m/z values.</p>
<p>To get a general overview of the full data we combine all spectra
measured for one sample into a single spectrum by reporting the maximum
intensity of peaks with highly similar m/z values across all spectra of
a sample using the <code>combineSpectra</code> function from the
<em>Spectra</em> package. Parameter <code>f</code> allows to define
which spectra in the data should be combined into a single spectrum.
Since we want to aggregate all spectra from one sample, we use
<code>f = fromFile(data)</code> (<code>fromFile</code> returns the
sample index for each individual spectrum in the data set).
<code>combineSpectra</code> uses the <code>combinePeaks</code> function
to aggregate signal for the defined groups of spectra. It provides a
large variety of possibilities to combine spectra, but in our case we
simply want to sum the signal of peaks with a very similar m/z and hence
we use parameters <code>intensityFun = max</code> and
<code>ppm = 10</code>. See also <code><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">?combinePeaks</a></code> for the full
set of parameters and aggregation options.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">combineSpectra</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, f <span class="op">=</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/Spectrum-class.html" class="external-link">fromFile</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>,</span>
<span>                      intensityFun <span class="op">=</span> <span class="va">max</span>, ppm <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectra</a></span><span class="op">(</span><span class="va">bps</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/unnamed-chunk-5-1.png" alt="Base peak spectrum for each of the two samples." width="672"><p class="caption">
Base peak spectrum for each of the two samples.
</p>
</div>
<p>In contrast to the single spectrum before, the base peak spectrum
(BPS) above shows more peaks indicating presence of more ions in the
samples for the m/z range.</p>
<p>In addition to a BPS we can also create a base peak chromatogram
(BPC) aggregating peak intensities for each scan (spectrum) per sample.
This BPC is orthogonal to the BPS and provides general information on
each LC run of the experiment. We use below the
<code>chromatogram</code> function to extract this data with the
additional parameter <code>aggregationFun = "max"</code> to report the
maximal intensity for each spectrum (and hence discrete retention
time).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bpc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bpc</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/unnamed-chunk-6-1.png" alt="Base peak chromatogram for the two samples. Each line represents one sample." width="672"><p class="caption">
Base peak chromatogram for the two samples. Each line represents one
sample.
</p>
</div>
<p>The BPC shows some noisy signal at the beginning of the
chromatography, but also some distinct (chromatographic) peak signal. In
addition, we can see slight drifts in retention time between the two
samples.</p>
<p>Apart from getting a general overview of the data it is also possible
to explore the data in more detail. To this end we next focus on a
specific subset of the data were we expect signal for a compound that
should be present in serum samples. We thus filter below the spectra
data using the <code>filterRt</code> function extracting only spectra
measured between 180 and 181 seconds.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">180</span>, <span class="fl">181</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sps</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 6 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##     msLevel     rtime scanIndex</span></span>
<span><span class="co">##   &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1         1   180.240       646</span></span>
<span><span class="co">## 2         1   180.519       647</span></span>
<span><span class="co">## 3         1   180.798       648</span></span>
<span><span class="co">## 4         1   180.235       646</span></span>
<span><span class="co">## 5         1   180.514       647</span></span>
<span><span class="co">## 6         1   180.793       648</span></span>
<span><span class="co">##  ... 33 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## 20171016_POOL_POS_1_105-134.mzML</span></span>
<span><span class="co">## 20171016_POOL_POS_3_105-134.mzML</span></span>
<span><span class="co">## Processing:</span></span>
<span><span class="co">##  Filter: select retention time [180..181] on MS level(s) 1 [Fri Sep  1 11:58:51 2023]</span></span></code></pre>
<p>For the present data set there were 6 spectra measured within this
one second in the two samples. By extracting the data as a
<code>Spectra</code> object we have however lost now the direct
(inherent) association between spectra and samples of the experiment. We
could extract the name of the original data file from which the data was
imported (see example below) and use that to determine the originating
sample, but that would involve additional R code.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">dataOrigin</a></span><span class="op">(</span><span class="va">sps</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "20171016_POOL_POS_1_105-134.mzML" "20171016_POOL_POS_1_105-134.mzML"</span></span>
<span><span class="co">## [3] "20171016_POOL_POS_1_105-134.mzML" "20171016_POOL_POS_3_105-134.mzML"</span></span>
<span><span class="co">## [5] "20171016_POOL_POS_3_105-134.mzML" "20171016_POOL_POS_3_105-134.mzML"</span></span></code></pre>
<p>Alternatively, we could use the <code>filterRt</code> function also
directly on the <code>MsExperiment</code> which would subset the
<code>MsExperiment</code> and hence the link between samples and spectra
would remain intact. Note however that only few filter and subset
functions are available for <code>MsExperiment</code> objects while a
large variety of useful filters are available for
<code>Spectra</code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' subset the whole MsExperiment</span></span>
<span><span class="va">data_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span><span class="va">data</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">180</span>, <span class="fl">181</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#' extract spectra from the subset for the first sample</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data_sub</span><span class="op">[</span><span class="fl">1L</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 3 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##     msLevel     rtime scanIndex</span></span>
<span><span class="co">##   &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1         1   180.240       646</span></span>
<span><span class="co">## 2         1   180.519       647</span></span>
<span><span class="co">## 3         1   180.798       648</span></span>
<span><span class="co">##  ... 33 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## 20171016_POOL_POS_1_105-134.mzML</span></span>
<span><span class="co">## Processing:</span></span>
<span><span class="co">##  Filter: select retention time [180..181] on MS level(s) 1 [Fri Sep  1 11:58:51 2023]</span></span></code></pre>
<p>For the present purpose it is however not important to keep the
sample association intact and we thus proceed to plot the previously
extracted spectra.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectra</a></span><span class="op">(</span><span class="va">sps</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/unnamed-chunk-10-1.png" alt="MS1 spectra measured between 180 and 181 seconds" width="672"><p class="caption">
MS1 spectra measured between 180 and 181 seconds
</p>
</div>
<p>We can immediately spot several mass peaks in the spectrum, with the
largest one at an m/z of about 130 and the second largest at about 106,
which could represent signal for an ion of <a href="https://en.wikipedia.org/wiki/Serine" class="external-link">Serine</a>. Below we
calculate the exact (monoisotopic) mass for serine from its chemical
formula <em>C3H7NO3</em> using the <code>calculateMass</code> function
from the <em><a href="https://bioconductor.org/packages/3.18/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></em>
package.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></span><span class="op">)</span></span>
<span><span class="va">mass_serine</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/calculateMass.html" class="external-link">calculateMass</a></span><span class="op">(</span><span class="st">"C3H7NO3"</span><span class="op">)</span></span>
<span><span class="va">mass_serine</span></span></code></pre></div>
<pre><code><span><span class="co">##  C3H7NO3 </span></span>
<span><span class="co">## 105.0426</span></span></code></pre>
<p>The <em>native</em> serine molecule is however uncharged and can thus
not be measured by mass spectrometry. In order to be detectable,
molecules need to be first ionized before being injected in an MS
instrument. While different ions can (and will) be generated for a
molecule, one of the most commonly created ions (adducts) in positive
polarity is the <em>[M+H]+</em> ion (protonated ion). To calculate the
m/z values for specific ions/adducts of molecules, we can use the
<code>mass2mz</code> function, also from the <em>MetaboCoreUtils</em>
package. Below we calculate the m/z for the <em>[M+H]+</em> ion of
serine providing the monoisotopic mass of that molecule and specifying
the ion we are interested in. Also other types of adducts are supported.
These could be listed with the <code>adductNames</code> function
(<code><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/adductNames.html" class="external-link">adductNames()</a></code> for all positively charged and
<code>adductNames("negative")</code> for all negatively charge
ions).</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">serine_mz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/mass2mz.html" class="external-link">mass2mz</a></span><span class="op">(</span><span class="va">mass_serine</span>, <span class="st">"[M+H]+"</span><span class="op">)</span></span>
<span><span class="va">serine_mz</span></span></code></pre></div>
<pre><code><span><span class="co">##           [M+H]+</span></span>
<span><span class="co">## C3H7NO3 106.0499</span></span></code></pre>
<p>The <code>mass2mz</code> function <strong>always</strong> returns a
<code>matrix</code> with columns reporting the m/z for the requested
adducts of the molecules (rows). Since we requested a single ion we
reduce this <code>matrix</code> to a single <code>numeric</code>.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">serine_mz</span> <span class="op">&lt;-</span> <span class="va">serine_mz</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<p>We can now use this information to subset the MS data to the signal
recorded for all ions with that particular m/z. We use again the
<code>chromatogram</code> function and provide the m/z range of interest
with the <code>mz</code> parameter.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">serine_chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, mz <span class="op">=</span> <span class="va">serine_mz</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.05</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">serine_chr</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/unnamed-chunk-14-1.png" alt="Ion trace for an ion of serine" width="672"><p class="caption">
Ion trace for an ion of serine
</p>
</div>
<p>A strong signal is visible around a retention time of 180 seconds
which very likely represents signal for the <em>[M+H]+</em> ion of
serine.</p>
<p>The object returned by the <code>chromatogram</code> function
arranges the individual <code>Chromatogram</code> objects (one per
sample) in a two-dimensional array, columns being samples (files) and
rows data slices (i.e. m/z - rt ranges). This type of data
representation is likely to be replaced in future with a more efficient
and flexible data structure similar to <code>Spectra</code>. Data from
the individual chromatograms can be accessed using the
<code>intensity</code> and <code>rtime</code> functions.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' get intensity and retention times for the chromatogram of the first</span></span>
<span><span class="co">#' sample</span></span>
<span><span class="va">ints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">intensity</a></span><span class="op">(</span><span class="va">serine_chr</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ints</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]  NA 559 659 278 492  NA</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">serine_chr</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.280 0.559 0.838 1.117 1.396 1.675</span></span></code></pre>
<p>At last we further focus on the tentative signal of serine extracting
the ion chromatogram restricting on a certain retention time range.
While we could also pass the retention time and m/z range with
parameters <code>rt</code> and <code>mz</code> to the
<code>chromatogram</code> function we instead filter the whole
experiment by retention time and m/z before calling
<code>chromatogram</code> on the such created data subset.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span>rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">175</span>, <span class="fl">189</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMz</a></span><span class="op">(</span>mz <span class="op">=</span> <span class="va">serine_mz</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.05</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/unnamed-chunk-15-1.png" alt="Extracted ion chromatogram for serine." width="672"><p class="caption">
Extracted ion chromatogram for serine.
</p>
</div>
<p>The area of such a chromatographic peak is supposed to be
proportional to the amount of the corresponding ion in the respective
sample and identification and quantification of such peaks is one of the
goals of the LC-MS data preprocessing.</p>
</div>
<div class="section level3">
<h3 id="centroiding-of-profile-ms-data">Centroiding of profile MS data<a class="anchor" aria-label="anchor" href="#centroiding-of-profile-ms-data"></a>
</h3>
<p>MS instruments allow to export data in profile or centroid mode.
Profile data contains the signal for all discrete m/z values (and
retention times) for which the instrument collected data <span class="citation">(R. Smith et al. 2014)</span>. MS instruments
continuously sample and record signals and a mass peak for a single ion
in one spectrum will thus consist of a multiple intensities at discrete
m/z values. Centroiding is the process to reduce these mass peaks to a
single representative signal, the centroid. This results in much smaller
file sizes, without loosing too much information. <em>xcms</em>,
specifically the <em>centWave</em> chromatographic peak detection
algorithm, was designed for centroided data, thus, prior to data
analysis, profile data, such as the example data used here, should be
centroided.</p>
<p>Below we inspect the profile data for the <em>[M+H]+</em> ion adduct
of Serine. We again subset the data to the m/z and retention time range
containing signal from Serine and use the <code>plot</code> function on
the subset <code>MsExperiment</code> to visualize the full MS data. This
<code>plot</code> function should ideally only called on subsets of an
<code>MsExperiment</code> but not on the full MS data. The plot
visualizes the intensities of individual peaks (color coded) in the
two-dimensional retention time against m/z space.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span>rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">175</span>, <span class="fl">189</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMz</a></span><span class="op">(</span>mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">106.02</span>, <span class="fl">106.07</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/serine-profile-mode-data-1.png" alt="Profile data for Serine." width="960"><p class="caption">
Profile data for Serine.
</p>
</div>
<p>The plot shows all data points measured by the instrument. Each
<em>column</em> of data points in the lower panel represents the signal
measured at one discrete time point, stored in one spectrum. We can see
a distribution of the signal for serine in both retention time and also
in m/z dimension.</p>
<p>Next we smooth the data in each spectrum using a Savitzky-Golay
filter, which usually improves data quality by reducing noise.
Subsequently we perform the centroiding of the data based on a simple
peak-picking strategy that reports the maximum signal for each mass peak
in each spectrum and replace the spectra data in our data
(<code>MsExperiment</code>) object.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Smooth and centroid the spectra data</span></span>
<span><span class="va">sps_cent</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"SavitzkyGolay"</span>, halfWindowSize <span class="op">=</span> <span class="fl">6L</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/hidden_aliases.html" class="external-link">pickPeaks</a></span><span class="op">(</span>halfWindowSize <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' replace spectra</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sps_cent</span></span>
<span></span>
<span><span class="co">#' Plot the centroided data for Serine</span></span>
<span><span class="va">data</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span>rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">175</span>, <span class="fl">189</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMz</a></span><span class="op">(</span>mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">106.02</span>, <span class="fl">106.07</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/centroiding-1.png" alt="Centroided data for Serine." width="960"><p class="caption">
Centroided data for Serine.
</p>
</div>
<p>The centroiding reduced the data to a single data point for an ion in
each spectrum. For more advanced centroiding options that can also
fine-tune the m/z value of the reported centroid see the
<code>pickPeaks</code> help or the centroiding vignette of the <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
package.</p>
<p>We next export the centroided MS data to files in mzML format and
re-read the data set from these.</p>
<p>We use the <code>export</code> function for data export of the
centroided <code>Spectra</code> object. Parameter <code>backend</code>
allows to specify the MS data backend that should be used for the
export, and that will also define the data format (use
<code>backend = MsBackendMzR()</code> to export data in mzML format.
Parameter <code>file</code> defines, for each spectrum, the name of the
file to which its data should be exported.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/hidden_aliases.html" class="external-link">export</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, backend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/MsBackend.html" class="external-link">MsBackendMzR</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>       file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">dataOrigin</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We next import the centroided data again from the newly generated
mzML files.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">fls</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Read the centroided data.</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/readMsExperiment.html" class="external-link">readMsExperiment</a></span><span class="op">(</span><span class="va">fls</span>, sampleData <span class="op">=</span> <span class="va">pd</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="preprocessing-of-lc-ms-data">Preprocessing of LC-MS data<a class="anchor" aria-label="anchor" href="#preprocessing-of-lc-ms-data"></a>
</h3>
<p>Preprocessing of (untargeted) LC-MS data aims at detecting and
quantifying the signal from ions generated from all molecules present in
a sample. It consists of the 3 steps chromatographic peak detection,
alignment (also called retention time correction) and correspondence
(also called peak grouping). The resulting matrix of feature abundances
can then be used as an input in downstream analyses including data
normalization, identification of features of interest and annotation of
features to metabolites.</p>
<div class="section level4">
<h4 id="chromatographic-peak-detection">Chromatographic peak detection<a class="anchor" aria-label="anchor" href="#chromatographic-peak-detection"></a>
</h4>
<p>Chromatographic peak detection aims to identify peaks along the
retention time axis that represent the signal from individual compounds’
ions. Such peak detection can be performed with the <em>xcms</em>
package using its <code>findChromPeaks</code> function. Several peak
detection algorithms are available that can be configured with its
specific parameter object: <code>MatchedFilterParam</code> to perform
peak detection as described in the original <em>xcms</em> article <span class="citation">(C. A. Smith et al. 2006)</span>,
<code>CentWaveParam</code> to perform a continuous wavelet
transformation (CWT)-based peak detection <span class="citation">(Tautenhahn, Böttcher, and Neumann 2008)</span> and
<code>MassifquantParam</code> to perform a Kalman filter-based peak
detection <span class="citation">(Conley et al. 2014)</span>. Additional
peak detection algorithms for direct injection data are also available,
but not discussed here.</p>
<p>In our example we use the <em>centWave</em> algorithm that performs
peak detection in two steps: first it identifies <em>regions of
interest</em> in the m/z - retention time space and subsequently detects
peaks in these regions using a continuous wavelet transform (see the
original publication for more details). The algorithm can be configured
with several parameters (see <code><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks-centWave.html" class="external-link">?CentWaveParam</a></code>), with the most
important being <code>peakwidth</code> and <code>ppm</code>.
<code>peakwidth</code> defines the minimal and maximal expected width of
the peak in retention time dimension and depends thus on the setup of
the employed LC-MS system making this parameter highly data set
dependent. Appropriate values can be estimated based on extracted ion
chromatograms of e.g. internal standards or known compounds in the data.
We thus below extract the chromatographic data for serine and perform a
peak detection on that data subset using <code>findChromPeaks</code> and
the default parameters for <em>centWave</em>.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Get the XIC for serine in all files</span></span>
<span><span class="va">serine_chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">164</span>, <span class="fl">200</span><span class="op">)</span>,</span>
<span>                           mz <span class="op">=</span> <span class="va">serine_mz</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.05</span>, <span class="fl">0.05</span><span class="op">)</span>,</span>
<span>                           aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Get default centWave parameters</span></span>
<span><span class="va">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks-centWave.html" class="external-link">CentWaveParam</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' "dry-run" peak detection on the XIC.</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks</a></span><span class="op">(</span><span class="va">serine_chr</span>, param <span class="op">=</span> <span class="va">cwp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      rt rtmin rtmax into intb maxo sn row column</span></span></code></pre>
<p>With the default settings for <em>centWave</em>,
<code>findChromPeaks</code> failed to detect any chromatographic peak in
that data. These default values are shown below.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cwp</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class:  CentWaveParam </span></span>
<span><span class="co">##  Parameters:</span></span>
<span><span class="co">##  - ppm: [1] 25</span></span>
<span><span class="co">##  - peakwidth: [1] 20 50</span></span>
<span><span class="co">##  - snthresh: [1] 10</span></span>
<span><span class="co">##  - prefilter: [1]   3 100</span></span>
<span><span class="co">##  - mzCenterFun: [1] "wMean"</span></span>
<span><span class="co">##  - integrate: [1] 1</span></span>
<span><span class="co">##  - mzdiff: [1] -0.001</span></span>
<span><span class="co">##  - fitgauss: [1] FALSE</span></span>
<span><span class="co">##  - noise: [1] 0</span></span>
<span><span class="co">##  - verboseColumns: [1] FALSE</span></span>
<span><span class="co">##  - roiList: list()</span></span>
<span><span class="co">##  - firstBaselineCheck: [1] TRUE</span></span>
<span><span class="co">##  - roiScales: numeric(0)</span></span>
<span><span class="co">##  - extendLengthMSW: [1] FALSE</span></span></code></pre>
<p>Particularly the settings for <code>peakwidth</code> does not fit our
data. The default for this parameter expects chromatographic peaks
between 20 and 50 seconds wide. When we plot the extracted ion
chromatogram (XIC) for serine we can however see that these values are
too large for the present data set (see below).</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">serine_chr</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/unnamed-chunk-17-1.png" alt="Extracted ion chromatogram for serine." width="672"><p class="caption">
Extracted ion chromatogram for serine.
</p>
</div>
<p>In fact, the serine signal seems to be measured for around 5 seconds.
We thus next adapt the settings to accommodate peaks ranging from 2 to
10 seconds and re-run the peak detection. In general, it is advised to
investigate peak widths for several ions in the data set to determine
the most appropriate <code>peakwidth</code> setting. In addition, we
select a different peak boundary estimation algorithm by setting
<code>integrate = 2</code>. This works particularly well for
non-gaussian peak shapes and ensures that also signal from peak’s tail
is integrated (eventually re-run the code with the default
<code>integrate = 1</code> to compare the two approaches).</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks-centWave.html" class="external-link">CentWaveParam</a></span><span class="op">(</span>peakwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">10</span><span class="op">)</span>, integrate <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">serine_chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks</a></span><span class="op">(</span><span class="va">serine_chr</span>, param <span class="op">=</span> <span class="va">cwp</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Plot the data and higlight identified peak area</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">serine_chr</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/centWave-adapted-1.png" alt="XIC for Serine with detected chromatographic peak" width="672"><p class="caption">
XIC for Serine with detected chromatographic peak
</p>
</div>
<p>With our data set-specific <code>peakwidth</code> we were able to
detect the peak for serine (highlighted in grey in the plot above). We
can now use the <code>chromPeaks</code> function to extract the
information on identified chromatographic peaks from our object.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">serine_chr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           rt   rtmin   rtmax     into     intb     maxo  sn row column</span></span>
<span><span class="co">## [1,] 181.356 178.566 189.447 74443.95 71734.01 37664.94 110   1      1</span></span>
<span><span class="co">## [2,] 181.072 178.561 187.210 70352.22 69008.98 38517.76 224   1      2</span></span></code></pre>
<p>The result is returned as a <code>matrix</code> with the retention
time and m/z range of the peak (<code>"rtmin"</code>,
<code>"rtmax"</code>, <code>"mzmin"</code> and <code>"mzmax"</code>) as
well as the integrated peak area (<code>"into"</code>), the maximal
signal (<code>"maxo"</code>) and the signal to noise ratio
(<code>"sn"</code>).</p>
<p>Another important parameter for <em>centWave</em> is <code>ppm</code>
which is used in the initial identification of the regions of interest.
In contrast to random noise, the <em>real</em> signal from an ion is
expected to yield stable m/z values in consecutive scans (the scattering
of the m/z values around the <em>real</em> m/z value of the ion is
supposed to be inversely related with its intensity - at least for TOF
instruments). In <em>centWave</em>, all peaks with m/z values that
differ by less than <code>ppm</code> in consecutive spectra are combined
into a <em>region of interest</em> (ROI) that is then subjected to the
CWT-based peak detection. To illustrate this, we plot the full MS data
for the data subset containing signal for serine.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Restrict the data to signal from Serine</span></span>
<span><span class="va">srn</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span>rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">179</span>, <span class="fl">186</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMz</a></span><span class="op">(</span>mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">106.04</span>, <span class="fl">106.07</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Plot the data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">srn</span><span class="op">)</span></span></code></pre></div>
<p><img src="xcms-preprocessing_files/figure-html/Serine-mz-scattering-plot-1.png" width="672"></p>
<p>We can observe some scattering of the data points in m/z dimension
(lower panel in the plot above), that decreases with increasing
intensity of the signal.</p>
<p>We next calculate the differences in m/z values between consecutive
scans in this data subset. We do this for one <em>representative</em>
file, selecting the one with the highest total sum of intensities in the
selected region. We calculate the total intensity by summing all
intensities of all spectra within the region for each of the two files
(the first <code>sum</code> on the intensities returned by
<code>intensity</code> sums all intensities within each spectrum, the
second <code>sum</code> sums these values to result in the total
sum).</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' split the experiment by sample</span></span>
<span><span class="va">srn_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">srn</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">srn</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' apply a function to each experiment that</span></span>
<span><span class="co">#' - extracts the spectra for that sample</span></span>
<span><span class="co">#' - gets their intensities</span></span>
<span><span class="co">#' - sums these intensities per spectrum</span></span>
<span><span class="co">#' - finally sums these intensity sums across spectra</span></span>
<span><span class="va">int_sums</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">srn_sample</span>,</span>
<span>                   <span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">intensity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">int_sums</span></span></code></pre></div>
<pre><code><span><span class="co">##        1        2 </span></span>
<span><span class="co">## 264511.6 251648.7</span></span></code></pre>
<p>We next extract the spectra from the sample with the highest
intensity and evaluate the number of peaks we’ve got for the m/z -
retention time range of our data subset.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">srn_sps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">srn</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">int_sums</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' get the number of peaks for each spectrum</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">srn_sps</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] 1 2 1 1 1 1 1 2 2 2 2 1 2 1 2 2 2 1 1 1 2 2 2 2 2</span></span></code></pre>
<p>For some spectra with have more than one peak in the MS data subset.
For these we next select the peak with the highest intensity. Here we
can make use of the powerful infrastructure of the <em>Spectra</em>
package, that allows to apply any user-provided function to the peak
matrix of each spectrum. We thus next define a simple function that
takes a peak matrix as input, subsets that to the one row (i.e. peak)
with the highest intensity and returns that single-row (peak) matrix
again.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' function to select the row with the highest intensity and return that</span></span>
<span><span class="va">max_int</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span>, <span class="st">"intensity"</span><span class="op">]</span><span class="op">)</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We can now apply this function to the data using the
<code>addProcessing</code> function. Any user provided function that
take a peak matrix as first argument and that return a peak matrix can
be passed to that <code>addProcessing</code> function. Note also that
the function has to have the <code>...</code> parameter in its function
definition (even if no additional parameters are passed along). See also
<code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">?addProcessing</a></code> for more information. Below we apply this
function to the extracted spectra and determine the number of peaks
after processing.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">srn_sps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">addProcessing</a></span><span class="op">(</span><span class="va">srn_sps</span>, <span class="va">max_int</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">srn_sps</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span></code></pre>
<p>We indeed have now a single peak per spectrum. We next extract the
m/z values for these peaks and calculate the difference of m/z values
between consecutive scans (spectra).</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mzs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">mz</a></span><span class="op">(</span><span class="va">srn_sps</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">mzs</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           mz           mz           mz           mz           mz           mz </span></span>
<span><span class="co">## 1.452460e-03 2.904891e-03 1.179878e-04 1.452442e-03 0.000000e+00 1.684509e-05 </span></span>
<span><span class="co">##           mz           mz           mz           mz           mz           mz </span></span>
<span><span class="co">## 0.000000e+00 0.000000e+00 7.233670e-05 0.000000e+00 0.000000e+00 7.624200e-07 </span></span>
<span><span class="co">##           mz           mz           mz           mz           mz           mz </span></span>
<span><span class="co">## 1.452441e-03 1.452441e-03 1.358206e-03 0.000000e+00 0.000000e+00 1.425717e-03 </span></span>
<span><span class="co">##           mz           mz           mz           mz           mz           mz </span></span>
<span><span class="co">## 0.000000e+00 1.452441e-03 1.480143e-03 0.000000e+00 0.000000e+00 1.493783e-03</span></span></code></pre>
<p>We can also express these differences in ppm (parts per million) of
the average m/z of the peaks.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">mzs</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1e6</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">mzs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           mz           mz           mz           mz           mz           mz </span></span>
<span><span class="co">## 13.695973646 27.391665930  1.112565444 13.695804399  0.000000000  0.158840806 </span></span>
<span><span class="co">##           mz           mz           mz           mz           mz           mz </span></span>
<span><span class="co">##  0.000000000  0.000000000  0.682098923  0.000000000  0.000000000  0.007189239 </span></span>
<span><span class="co">##           mz           mz           mz           mz           mz           mz </span></span>
<span><span class="co">## 13.695795336 13.695795336 12.807200180  0.000000000  0.000000000 13.443799681 </span></span>
<span><span class="co">##           mz           mz           mz           mz           mz           mz </span></span>
<span><span class="co">##  0.000000000 13.695795190 13.957010392  0.000000000  0.000000000 14.085629933</span></span></code></pre>
<p>The difference in m/z values for the serine data is thus between 0
and 27 ppm. This should ideally be evaluated for several compounds and
should be set to a value that allows to capture the full chromatographic
peaks for most of the tested compounds. We can next perform the peak
detection using our settings for the <code>ppm</code> and
<code>peakwidth</code> parameters.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Perform peak detection</span></span>
<span><span class="va">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks-centWave.html" class="external-link">CentWaveParam</a></span><span class="op">(</span>peakwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">10</span><span class="op">)</span>, ppm <span class="op">=</span> <span class="fl">30</span>, integrate <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks</a></span><span class="op">(</span><span class="va">data</span>, param <span class="op">=</span> <span class="va">cwp</span><span class="op">)</span></span></code></pre></div>
<p>The <code>findChromPeaks</code> call adds the results from the
chromatographic peak detection to our data set (which is now represented
by an <code>XcmsExperiment</code> object that directly extends
<code>MsExperiment</code> and hence inherits all of its
functionality).</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class XcmsExperiment </span></span>
<span><span class="co">##  Spectra: MS1 (1862) </span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - spectra: 2 sample(s) to 1862 element(s).</span></span>
<span><span class="co">##  xcms results:</span></span>
<span><span class="co">##   - chromatographic peaks: 653 in MS level(s): 1</span></span></code></pre>
<p>The results from the peak detection analysis can be accessed with the
<code>chromPeaks</code> function. The optional parameters
<code>rt</code> and <code>mz</code> allow in addition to extract peak
detection results for a specific m/z - retention time region:</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Access the peak detection results from a specific m/z - rt area</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">data</span>, mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">106</span>, <span class="fl">107</span><span class="op">)</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">150</span>, <span class="fl">190</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             mz    mzmin    mzmax      rt   rtmin   rtmax       into       intb</span></span>
<span><span class="co">## CP133 106.0625 106.0606 106.0636 173.264 171.869 174.380   516.3588   510.5323</span></span>
<span><span class="co">## CP167 106.0506 106.0505 106.0506 181.356 178.845 187.773 74181.7823 73916.8683</span></span>
<span><span class="co">## CP475 106.0633 106.0609 106.0652 172.701 170.748 174.654   559.5491   553.2785</span></span>
<span><span class="co">## CP516 106.0496 106.0494 106.0508 181.072 178.282 187.210 70373.6099 70106.7152</span></span>
<span><span class="co">##             maxo  sn sample</span></span>
<span><span class="co">## CP133   426.6084  38      1</span></span>
<span><span class="co">## CP167 37664.9371 688      1</span></span>
<span><span class="co">## CP475   381.6084  53      2</span></span>
<span><span class="co">## CP516 38517.7622 826      2</span></span></code></pre>
<p>For each identified peak the m/z and rt value of the apex is reported
(columns <code>"mz"</code> and <code>"rt"</code>) as well as their
ranges (<code>"mzmin"</code>, <code>"mzmax"</code>,
<code>"rtmin"</code>, <code>"rtmax"</code>), the integrated signal of
the peak (i.e. the peak area <code>"into"</code>), the maximal signal of
the peak (<code>"maxo"</code>), the signal to noise ratio
(<code>"sn"</code>) and the index of the sample in which the peak was
detected (<code>"sample"</code>).</p>
<p>Note that <strong>after</strong> peak detection, extracted ion
chromatogram data will also contain identified chromatographic peaks.
Below we extract the EIC for Serine and display the identified peaks for
that compound.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eic_serine</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">106.04</span>, <span class="fl">106.06</span><span class="op">)</span>,</span>
<span>                           rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">179</span>, <span class="fl">186</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_serine</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             mz    mzmin    mzmax      rt   rtmin   rtmax     into     intb</span></span>
<span><span class="co">## CP167 106.0506 106.0505 106.0506 181.356 178.845 187.773 74181.78 73916.87</span></span>
<span><span class="co">## CP516 106.0496 106.0494 106.0508 181.072 178.282 187.210 70373.61 70106.72</span></span>
<span><span class="co">##           maxo  sn sample row column</span></span>
<span><span class="co">## CP167 37664.94 688      1   1      1</span></span>
<span><span class="co">## CP516 38517.76 826      2   1      2</span></span></code></pre>
<p>And plotting this extracted ion chromatogram will also show the
identified chromatographic peak(s).</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">eic_serine</span><span class="op">)</span></span></code></pre></div>
<p><img src="xcms-preprocessing_files/figure-html/plot-after-1.png" width="672"></p>
<p>This thus provides a convenient way to evaluate peak detection
results for sets of m/z - retention time regions for potential known
compounds.</p>
<p>Similarly, peak detection results will be visualized in the generic
plot of the result object (chromatogrphic peaks are highlighed as a red
rectangle; see below).</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">srn</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span>rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">175</span>, <span class="fl">188</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMz</a></span><span class="op">(</span>mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">106.04</span>, <span class="fl">106.06</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">srn</span><span class="op">)</span></span></code></pre></div>
<p><img src="xcms-preprocessing_files/figure-html/xic-after-1.png" width="672"></p>
<p>Peak detection will not always work perfectly leading to peak
detection artifacts, such as overlapping peaks or artificially split
peaks. The <code>refineChromPeaks</code> function allows to
<em>refine</em> peak detection results by either removing peaks not
meeting predefined properties or by merging artificially split
chromatographic peaks (see <code><a href="https://rdrr.io/pkg/xcms/man/refineChromPeaks.html" class="external-link">?refineChromPeaks</a></code> for all
options). Below we post-process the peak detection results merging peaks
(within each sample) that overlap in a 4 second window if the signal
between in between them is lower than 75% of the smaller peak’s largest
intensity. See the <code>MergeNeighboringPeaksParam</code> help page for
a detailed description of the settings and the approach.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mpp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/refineChromPeaks.html" class="external-link">MergeNeighboringPeaksParam</a></span><span class="op">(</span>expandRt <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">data_pp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/refineChromPeaks.html" class="external-link">refineChromPeaks</a></span><span class="op">(</span><span class="va">data</span>, param <span class="op">=</span> <span class="va">mpp</span><span class="op">)</span></span></code></pre></div>
<p>Merged peaks can be identified with a <code>TRUE</code> in the
<code>"merged"</code> chromatographic peak metadata column that can be
extracted with the <code>chromPeakData</code> function (see below).</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeakData</a></span><span class="op">(</span><span class="va">data_pp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 595 rows and 3 columns</span></span>
<span><span class="co">##        ms_level is_filled    merged</span></span>
<span><span class="co">##       &lt;integer&gt; &lt;logical&gt; &lt;logical&gt;</span></span>
<span><span class="co">## CP001         1     FALSE     FALSE</span></span>
<span><span class="co">## CP002         1     FALSE     FALSE</span></span>
<span><span class="co">## CP003         1     FALSE     FALSE</span></span>
<span><span class="co">## CP004         1     FALSE     FALSE</span></span>
<span><span class="co">## CP005         1     FALSE     FALSE</span></span>
<span><span class="co">## ...         ...       ...       ...</span></span>
<span><span class="co">## CP679         1     FALSE      TRUE</span></span>
<span><span class="co">## CP680         1     FALSE      TRUE</span></span>
<span><span class="co">## CP681         1     FALSE      TRUE</span></span>
<span><span class="co">## CP682         1     FALSE      TRUE</span></span>
<span><span class="co">## CP683         1     FALSE      TRUE</span></span></code></pre>
<p>An example for a joined (merged) peak is given below.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mzr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">122.9</span>, <span class="fl">122.97</span><span class="op">)</span></span>
<span><span class="va">rtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">150</span><span class="op">)</span></span>
<span></span>
<span><span class="va">chr_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, mz <span class="op">=</span> <span class="va">mzr</span>, rt <span class="op">=</span> <span class="va">rtr</span><span class="op">)</span></span>
<span><span class="va">chr_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data_pp</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, mz <span class="op">=</span> <span class="va">mzr</span>, rt <span class="op">=</span> <span class="va">rtr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_2</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/merged-peak-1.png" alt="Result from the peak refinement. Left: peaks before merging, right after merging." width="1152"><p class="caption">
Result from the peak refinement. Left: peaks before merging, right after
merging.
</p>
</div>
<p>The signal is rather wide and noise and <em>centWave</em> thus was
not able to integrate the whole peak and even defined partially
overlapping chromatographic peaks (left in the image above). Peak
refinement merged all the consecutive peaks as the signal between them
never dropped below 75% of the smallest chromatographic peak’s
intensity.</p>
<p>At last we replace the <code>data</code> variable with the object
containing also the peak refinement results.</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data_pp</span></span></code></pre></div>
<p>For quality assessment we could now calculate summary statistics on
the identified peaks to e.g. identify samples with much less detected
peaks. Also, we can use the <code>plotChromPeaks</code> function to
provide some general information on the location of the identified
chromatographic peaks in the m/z - rt space.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeaks.html" class="external-link">plotChromPeaks</a></span><span class="op">(</span><span class="va">data</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeaks.html" class="external-link">plotChromPeaks</a></span><span class="op">(</span><span class="va">data</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/plotChromPeaks-1.png" alt="Location of the identified chromatographic peaks in the m/z - rt space." width="1152"><p class="caption">
Location of the identified chromatographic peaks in the m/z - rt space.
</p>
</div>
</div>
<div class="section level4">
<h4 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h4>
<p>While chromatography helps to discriminate better between analytes it
is also affected by variances that can lead to shifts in retention times
between measurement runs. The alignment step aims to adjust these
retention time differences between samples within an experiment. Below
we plot the base peak chromatograms of both files of our data set to
visualize these differences. Note that with
<code>peakType = "none"</code> we disable plotting of identified
chromatographic peaks that would be drawn by default on chromatograms
extracted from an object containing peak detection results.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extract base peak chromatograms</span></span>
<span><span class="va">bpc_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bpc_raw</span>, peakType <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/alignment-bpc-raw-1.png" alt="BPC of all files." width="768"><p class="caption">
BPC of all files.
</p>
</div>
<p>While both samples were measured with the same setup on the same day
slight drifts of the signal are visible.</p>
<p>Alignment can be performed with <em>xcms</em> using the
<code>adjustRtime</code> function that supports the <em>peakGroups</em>
<span class="citation">(C. A. Smith et al. 2006)</span> and the
<em>obiwarp</em> <span class="citation">(Prince and Marcotte
2006)</span> method. The settings for the algorithms can be defined with
the <code>PeakGroupsParam</code> and the <code>ObiwarpParam</code>
parameter objects, respectively (see <code><a href="https://rdrr.io/pkg/xcms/man/adjustRtime.html" class="external-link">?adjustRtime</a></code> for all
available alignment methods and their settings). Note also that
<em>xcms</em> supports a <em>subset-based</em> alignment which allows to
align a large data set with the retention time drift estimated on a
subset of the samples (ideally QC or pooled samples repeatedly measured
over the whole measurement run(s)). See the <em>xcms</em> <a href="https://bioconductor.org/packages/release/bioc/vignettes/xcms/inst/doc/xcms.html" class="external-link">vignette</a>
for more information.</p>
<p>For our example we use the <em>peakGroups</em> method that aligns
samples based on the retention times of <em>hook peaks</em> (or
housekeeping peaks), which represent signal from ions expected to be
present in most samples. To define these we need however to group
detected chromatographic peaks across samples using an initial
correspondence analysis. Below we use the <em>peakDensity</em> method
for correspondence. Details about this method and explanations on the
choices of its parameters are provided in the next section. After having
performed this initial correspondence analysis, we perform the alignment
using settings <code>minFraction = 1</code> and <code>span = 0.6</code>.
<code>minFraction</code> defines the proportion of samples in which a
candidate hook peak has to be detected/present. A value of 0.9 would for
example require the chromatographic peak for a specific ion (i.e. m/z -
retention time) to be present (detected) in 90% of samples to consider
it as a hook peak for the alignment. Our data represents replicated
measurements of the same sample pool and we can therefore require hook
peaks to be present in each file. The parameter <code>span</code>
defines the degree of smoothing of the loess function that is used to
allow different regions along the retention time axis to be adjusted by
a different factor. A value of 0 will most likely cause overfitting,
while 1 would perform a constant, linear shift. Values between 0.4 and
0.6 seem to be reasonable for most experiments.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Define the settings for the initial peak grouping - details for</span></span>
<span><span class="co">#' choices in the next section.</span></span>
<span><span class="va">pdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">$</span><span class="va">group</span>, bw <span class="op">=</span> <span class="fl">1.8</span>,</span>
<span>                        minFraction <span class="op">=</span> <span class="fl">1</span>, binSize <span class="op">=</span> <span class="fl">0.02</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">groupChromPeaks</a></span><span class="op">(</span><span class="va">data</span>, <span class="va">pdp</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Define settings for the alignment</span></span>
<span><span class="va">pgp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/adjustRtime.html" class="external-link">PeakGroupsParam</a></span><span class="op">(</span>minFraction <span class="op">=</span> <span class="fl">1</span>, span <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/adjustRtime.html" class="external-link">adjustRtime</a></span><span class="op">(</span><span class="va">data</span>, param <span class="op">=</span> <span class="va">pgp</span><span class="op">)</span></span></code></pre></div>
<p>Adjusted retention times are stored, along with the raw retention
times, within the result object. Any function accessing retention times
(such as <code>rtime</code>) will by default return adjusted retention
times from an <code>XcmsExperiment</code> object, if present. Note that
also the retention times of the identified chromatographic peaks were
adjusted by the <code>adjustRtime</code> call. After alignment it is
suggested to evaluate alignment results e.g. by inspecting differences
between raw and adjusted retention times.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Plot the difference between raw and adjusted retention times</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotAdjustedRtime.html" class="external-link">plotAdjustedRtime</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/alignment-result-1.png" alt="Alignment results. Shown is the difference between raw and adjusted retention times and the hook peaks that were used for the alignment (shown as points)." width="768"><p class="caption">
Alignment results. Shown is the difference between raw and adjusted
retention times and the hook peaks that were used for the alignment
(shown as points).
</p>
</div>
<p>The difference between raw and adjusted retention time should be
reasonable. In our example it is mostly below one second, which is OK
since the samples were measured within a short time period and
differences are thus expected to be small. Also, hook peaks should
ideally be present along the full retention time range to allow proper
estimation of drifts along the full time of measurement.</p>
<p>Next we plot the base peak chromatograms before and after alignment.
Note that a <code>chromatogram</code> call will always include all
detected chromatographic peaks in the requested rt range but for a base
peak chromatogram we don’t need this information. With
<code>chromPeaks = "none"</code> in the <code>chromatogram</code> call
we tell the function to <strong>not</strong> include any identified
chromatographic peaks in the returned chromatographic data.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#' Plot the raw base peak chromatogram</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bpc_raw</span>, peakType <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="co">#' Plot the BPC after alignment</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span>, chromPeaks <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/bpc-raw-adjusted-1.png" alt="BPC before (top) and after (bottom) alignment." width="960"><p class="caption">
BPC before (top) and after (bottom) alignment.
</p>
</div>
<p>The base peak chromatograms are nicely aligned after retention time
adjustment. The impact of the alignment should also be evaluated on
known compounds or internal standards. We thus plot below the XIC for
serine before and after alignment. To get raw retention times and hence
be able to extract the ion chromatogram of serine before alignment we
have to <em>drop</em> the alignment results which can be done with the
<code>dropAdjustedRtime</code> function. This function will restore the
retention times of spectra (and chromatographic peaks)
<strong>before</strong> alignment.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">dropAdjustedRtime</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<p>We can thus extract next the ion chromatogram for serine from the raw
data as well as after alignment.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Use adjustedRtime parameter to access raw/adjusted retention times</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4.5</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data_raw</span>, mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">106.04</span>, <span class="fl">106.06</span><span class="op">)</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">179</span>, <span class="fl">186</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">106.04</span>, <span class="fl">106.06</span><span class="op">)</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">179</span>, <span class="fl">186</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/serine-xic-adjusted-1.png" alt="XIC for Serine before (left) and after (right) alignment" width="960"><p class="caption">
XIC for Serine before (left) and after (right) alignment
</p>
</div>
<p>The Serine peaks are also nicely aligned after adjustment.</p>
</div>
<div class="section level4">
<h4 id="correspondence">Correspondence<a class="anchor" aria-label="anchor" href="#correspondence"></a>
</h4>
<p>The final step of the LC-MS preprocessing with <em>xcms</em> is the
correspondence analysis, in which chromatographic peaks from the same
types of ions (compounds) are grouped across samples to form a so called
<em>LC-MS feature</em>. <code>xcms</code> implements two methods for
this purpose: <em>peak density</em> <span class="citation">(C. A. Smith
et al. 2006)</span> and <em>nearest</em> <span class="citation">(Katajamaa, Miettinen, and Oresic 2006)</span> that can
be configured by passing either a <code>PeakDensityParam</code> or a
<code>NearestPeaksParam</code> object to the
<code>groupChromPeaks</code> function (see <code><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">?groupChromPeaks</a></code>
for more information). For our example we use the <em>peak density</em>
method that iterates through m/z slices in the data and groups
chromatographic peaks to features in each slice (within the same or
across samples) depending on their retention time and the distribution
of chromatographic peaks along the retention time axis. Peaks
representing signal from the same ion are expected to have a similar
retention time and, if found in many samples, this should also be
reflected by a higher peak density at the respective retention time
(since the retention times for this compound are expected to be similar
in the different samples). To illustrate this we extract below an m/z
slice containing the serine peak and use the
<code>plotChromPeakDensity</code> function to visualize the distribution
of peaks along the retention time axis and to <em>simulate</em> a
correspondence analysis based on the provided settings.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Get default parameters for the grouping</span></span>
<span><span class="va">pdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Extract a BPC for the m/z slice containing serine</span></span>
<span><span class="va">bpc_serine</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">106.04</span>, <span class="fl">106.06</span><span class="op">)</span>,</span>
<span>                           aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Dry-run correspondence and show the results.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity</a></span><span class="op">(</span><span class="va">bpc_serine</span>, param <span class="op">=</span> <span class="va">pdp</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/correspondence-example-1.png" alt="BPC for a m/z slice and defined features within this slice based on default settings." width="960"><p class="caption">
BPC for a m/z slice and defined features within this slice based on
default settings.
</p>
</div>
<p>The upper panel in the plot above shows the chromatographic data with
the identified peaks. The lower panel shows the retention time of
identified peaks (x-axis) per sample (y-axis) with the black solid line
representing their distribution (density) along the x-axis. Peak groups
(features) are indicated with grey rectangles in the lower panel. The
peak density correspondence method groups all chromatographic peaks
under the same <em>density peak</em> into one feature. With the default
settings we were able to group the serine peak of each sample into one
feature. The parameters for the peak density correspondence analysis
are:</p>
<ul>
<li>
<code>binSize</code>: m/z width of the bin/slice of data in which
peaks are grouped.</li>
<li>
<code>bw</code> defines the smoothness of the density function.</li>
<li>
<code>maxFeatures</code>: maximum number of features to be defined
in one bin.</li>
<li>
<code>minFraction</code>: minimum proportion of samples (of one
group!) for which a peak has to be present.</li>
<li>
<code>minSamples</code>: minimum number of samples a peak has to be
present.</li>
</ul>
<p>The parameters <code>minFraction</code> and <code>minSamples</code>
depend on the experimental layout and should be set accordingly.
<code>binSize</code> should be set to a small enough value to avoid
peaks from different ions, but with similar m/z and retention time,
being grouped together. The most important parameter however is
<code>bw</code> and, while its default value of 30 was able to correctly
group the Serine peaks, it should always be evaluated on other, more
complicated, signals too. Below we evaluate the performance of the
default parameters on an m/z slice that contains signal from multiple
ions with the same m/z, including isomers betaine and valine ([M+H]+ m/z
118.08625).</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Plot the chromatogram for an m/z slice containing Betaine and Valine</span></span>
<span><span class="va">mzr</span> <span class="op">&lt;-</span> <span class="fl">118.08625</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.01</span>, <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="va">chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, mz <span class="op">=</span> <span class="va">mzr</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Correspondence in that slice using default settings</span></span>
<span><span class="va">pdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity</a></span><span class="op">(</span><span class="va">chr</span>, param <span class="op">=</span> <span class="va">pdp</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/correspondence-bw-1.png" alt="Correspondence analysis with default settings on an m/z slice containing signal from multiple ions." width="960"><p class="caption">
Correspondence analysis with default settings on an m/z slice containing
signal from multiple ions.
</p>
</div>
<p>With default settings all chromatographic peaks present in the m/z
slice were grouped into one feature. Signal from different ions would
thus be treated as a single entity. Below we repeat the analysis with a
strongly reduced value for <code>bw</code>.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Reducing the bandwidth</span></span>
<span><span class="va">pdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">$</span><span class="va">group</span>, bw <span class="op">=</span> <span class="fl">1.8</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity</a></span><span class="op">(</span><span class="va">chr</span>, param <span class="op">=</span> <span class="va">pdp</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/correspondence-bw-fix-1.png" alt="Correspondence analysis with reduced bw setting on a m/z slice containing signal from multiple ions." width="960"><p class="caption">
Correspondence analysis with reduced bw setting on a m/z slice
containing signal from multiple ions.
</p>
</div>
<p>Using a value of <code>1.8</code> for parameter <code>bw</code>, we
successfully grouped the peaks into different features. We can now use
these settings for the correspondence analysis on the full data set.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">$</span><span class="va">group</span>, bw <span class="op">=</span> <span class="fl">1.8</span>,</span>
<span>                        minFraction <span class="op">=</span> <span class="fl">0.4</span>, binSize <span class="op">=</span> <span class="fl">0.02</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Perform the correspondence analysis</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">groupChromPeaks</a></span><span class="op">(</span><span class="va">data</span>, param <span class="op">=</span> <span class="va">pdp</span><span class="op">)</span></span></code></pre></div>
<p>Next we evaluate the results from the correspondence analysis on a
different m/z slice containing isomers leucine and isoleucine ([M+H]+
m/z 132.10191). Setting <code>simulate = FALSE</code> in
<code>plotChromPeakDensity</code> will show the actual results from the
correspondence analysis.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Plot the results for an m/z slice containing Leucine and Isoleucine</span></span>
<span><span class="va">mzr</span> <span class="op">&lt;-</span> <span class="fl">132.10191</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.01</span>, <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="va">chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, mz <span class="op">=</span> <span class="va">mzr</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity</a></span><span class="op">(</span><span class="va">chr</span>, simulate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-preprocessing_files/figure-html/correspondence-evaluate-1.png" alt="Result of correspondence on a slice containing the isomers Leucine and Isoleucine." width="960"><p class="caption">
Result of correspondence on a slice containing the isomers Leucine and
Isoleucine.
</p>
</div>
<p>Despite being very close, chromatographic peaks of isomers were
successfully grouped into separate features.</p>
<p>Results from the correspondence analysis can be accessed with the
<code>featureDefinition</code> function. This function returns a data
frame with the retention time and m/z ranges of the apex positions from
the peaks assigned to the feature and their respective indices in the
<code>chromPeaks</code> matrix.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Definition of the features</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">featureDefinitions</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          mzmed    mzmin    mzmax     rtmed     rtmin     rtmax npeaks POOL</span></span>
<span><span class="co">## FT001 105.0418 105.0417 105.0418 167.68597 167.48455 167.88740      2    2</span></span>
<span><span class="co">## FT002 105.0415 105.0415 105.0415 157.71871 157.71871 157.71871      1    1</span></span>
<span><span class="co">## FT003 105.0697 105.0691 105.0703  31.80794  31.68918  31.92670      2    2</span></span>
<span><span class="co">## FT004 105.1103 105.1100 105.1105  63.75047  63.35239  64.14855      2    2</span></span>
<span><span class="co">## FT005 105.4734 105.4732 105.4736 201.57593 201.36133 201.79053      2    2</span></span>
<span><span class="co">## FT006 105.7166 105.7160 105.7172 181.21578 181.08901 181.34256      2    2</span></span>
<span><span class="co">##        peakidx ms_level</span></span>
<span><span class="co">## FT001 112, 396        1</span></span>
<span><span class="co">## FT002      111        1</span></span>
<span><span class="co">## FT003  19, 317        1</span></span>
<span><span class="co">## FT004  48, 348        1</span></span>
<span><span class="co">## FT005 260, 580        1</span></span>
<span><span class="co">## FT006 135, 444        1</span></span></code></pre>
<p>Note that the EIC <code>chr</code> above would also contain the
correspondence results for the selected m/z range which could also be
accessed with <code>featureDefinitions(chr)</code>.</p>
<p>Also, we can calculate simple per-feature summary statistic with the
<code>featureSummary</code> function. This function reports for each
feature the total number and the percentage of samples in which a peak
was detected and the total numbers and percentage of these samples in
which more than one peak was assigned to the feature.</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Per-feature summary.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/featureSummary.html" class="external-link">featureSummary</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       count perc multi_count multi_perc        rsd</span></span>
<span><span class="co">## FT001     2  100           0          0 0.23642129</span></span>
<span><span class="co">## FT002     1   50           0          0         NA</span></span>
<span><span class="co">## FT003     2  100           0          0 0.24525296</span></span>
<span><span class="co">## FT004     2  100           0          0 0.04687951</span></span>
<span><span class="co">## FT005     2  100           0          0 0.22330558</span></span>
<span><span class="co">## FT006     2  100           0          0 0.05904323</span></span></code></pre>
<p>The final result from the LC-MS data preprocessing is a matrix with
feature abundances, rows being features, columns samples. Such a matrix
can be extracted with the <code>featureValues</code> function from the
result object (see further below for an alternative, preferred way to
extract preprocessing results with the <code>quantify</code> method).
The function takes two additional parameters <code>value</code> and
<code>method</code>: <code>value</code> defines the column in the
<code>chromPeaks</code> table that should be reported in the matrix, and
<code>method</code> the approach to handle cases in which more than one
peak in a sample is assigned to the feature.</p>
<p>Below we set <code>value = "into"</code> (the default) to extract the
total integrated peak area and <code>method = "maxint"</code> to report
the peak area of the peak with the largest intensity for features with
multiple peaks in a sample.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' feature intensity matrix</span></span>
<span><span class="va">fmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">data</span>, value <span class="op">=</span> <span class="st">"into"</span>, method <span class="op">=</span> <span class="st">"maxint"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fmat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       20171016_POOL_POS_1_105-134.mzML 20171016_POOL_POS_3_105-134.mzML</span></span>
<span><span class="co">## FT001                        3202.7445                        2285.2830</span></span>
<span><span class="co">## FT002                        3605.3915                               NA</span></span>
<span><span class="co">## FT003                         744.8752                        1057.4312</span></span>
<span><span class="co">## FT004                       18126.4603                       19369.4039</span></span>
<span><span class="co">## FT005                       23243.6129                       31960.3709</span></span>
<span><span class="co">## FT006                         671.5842                         617.7545</span></span></code></pre>
<p>While we do have abundances reported for most features, we might also
have missing values for some, like for feature <em>FT002</em> in the
second sample above. Such <code>NA</code>s occur if no chromatographic
peak was assigned to a feature, either because peak detection failed, or
because the corresponding ion is absent in the respective sample. One
possibility to deal with such missing values is data imputation. With
the <code>fillChromPeaks</code> function, <em>xcms</em> provides however
an alternative approach that integrates the signal measured at the m/z -
retention time region of the feature in the original files of samples
for which an <code>NA</code> was reported hence <em>filling-in</em>
missing peak data. Different approaches for this gap-filling are
available (see <code>?fillChromPeals</code>), with the method selected
and configured with <code>ChromPeakAreaParam</code> being the preferred
one: this approach defines the region from which the signal should be
integrated based on the m/z and rt range of the identified
chromatographic peaks of the feature.</p>
<p>Below we perform the gap-filling with the
<code>ChromPeakAreaParam</code> approach.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Number of missing values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">fmat</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 135</span></span></code></pre>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/fillChromPeaks.html" class="external-link">fillChromPeaks</a></span><span class="op">(</span><span class="va">data</span>, param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/fillChromPeaks.html" class="external-link">ChromPeakAreaParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' How many missing values after</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 28</span></span></code></pre>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fmat_fld</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">data</span>, value <span class="op">=</span> <span class="st">"into"</span>, method <span class="op">=</span> <span class="st">"maxint"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fmat_fld</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       20171016_POOL_POS_1_105-134.mzML 20171016_POOL_POS_3_105-134.mzML</span></span>
<span><span class="co">## FT001                        3202.7445                        2285.2830</span></span>
<span><span class="co">## FT002                        3605.3915                        3183.9546</span></span>
<span><span class="co">## FT003                         744.8752                        1057.4312</span></span>
<span><span class="co">## FT004                       18126.4603                       19369.4039</span></span>
<span><span class="co">## FT005                       23243.6129                       31960.3709</span></span>
<span><span class="co">## FT006                         671.5842                         617.7545</span></span></code></pre>
<p>With <code>fillChromPeaks</code> we could thus <em>rescue</em> signal
for all but 23 features with missing values. Note that filled-in peak
information can also be removed any time with the
<code>dropFilledChromPeaks</code> function. Also, setting
<code>filled = FALSE</code> in the <code>featureValues</code> function
would return only data from detected peaks.</p>
<p>Note also that, in addition to the <code>featureValues</code> method
that simply extracts the feature matrix, the preprocessing results can
also be converted to a <code>SummarizedExperiment</code> that, besides
containing the feature quantification, stores also the phenotype data
(sample descriptions) and feature definitions. Below we use the
<code>quantify</code> method to extract the preprocessing results as a
<code>SummarizedExperiment</code>. The function takes any parameter of
the <code>featureValues</code> function as additional arguments, thus,
with the call below we extract all values (from detected and gap-filled
peaks) and sum up the signals of all chromatographic peaks assigned to a
feature in a sample.</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">quantify</a></span><span class="op">(</span><span class="va">data</span>, method <span class="op">=</span> <span class="st">"sum"</span>, value <span class="op">=</span> <span class="st">"into"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>SummarizedExperiment</code> is the main data object in
Bioconductor to store quantified omics data and hence an ideal container
for the LC-MS data preprocessing results. The column annotations can be
accessed with <code>colData</code>, the feature definitions (row
annotations) with <code>rowData</code>:</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2 rows and 5 columns</span></span>
<span><span class="co">##                                           file injection_idx      sample</span></span>
<span><span class="co">##                                    &lt;character&gt;     &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">## 20171016_POOL_POS_1_105-134.mzML 20171016_P...             1      POOL_1</span></span>
<span><span class="co">## 20171016_POOL_POS_3_105-134.mzML 20171016_P...            19      POOL_2</span></span>
<span><span class="co">##                                        group spectraOrigin</span></span>
<span><span class="co">##                                  &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">## 20171016_POOL_POS_1_105-134.mzML        POOL /__w/xcmsT...</span></span>
<span><span class="co">## 20171016_POOL_POS_3_105-134.mzML        POOL /__w/xcmsT...</span></span></code></pre>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 361 rows and 9 columns</span></span>
<span><span class="co">##           mzmed     mzmin     mzmax     rtmed     rtmin     rtmax    npeaks</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## FT001   105.042   105.042   105.042  167.6860  167.4845  167.8874         2</span></span>
<span><span class="co">## FT002   105.042   105.042   105.042  157.7187  157.7187  157.7187         1</span></span>
<span><span class="co">## FT003   105.070   105.069   105.070   31.8079   31.6892   31.9267         2</span></span>
<span><span class="co">## FT004   105.110   105.110   105.111   63.7505   63.3524   64.1486         2</span></span>
<span><span class="co">## FT005   105.473   105.473   105.474  201.5759  201.3613  201.7905         2</span></span>
<span><span class="co">## ...         ...       ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">## FT357   133.928   133.928   133.928  198.5823  198.5823  198.5823         1</span></span>
<span><span class="co">## FT358   133.960   133.960   133.961   30.8309   30.8063   30.8554         2</span></span>
<span><span class="co">## FT359   133.956   133.956   133.956  199.0661  198.9950  199.1372         2</span></span>
<span><span class="co">## FT360   133.973   133.973   133.973  206.8712  206.3899  207.3524         2</span></span>
<span><span class="co">## FT361   133.973   133.973   133.973  200.2484  200.2484  200.2484         1</span></span>
<span><span class="co">##            POOL  ms_level</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## FT001         2         1</span></span>
<span><span class="co">## FT002         1         1</span></span>
<span><span class="co">## FT003         2         1</span></span>
<span><span class="co">## FT004         2         1</span></span>
<span><span class="co">## FT005         2         1</span></span>
<span><span class="co">## ...         ...       ...</span></span>
<span><span class="co">## FT357         1         1</span></span>
<span><span class="co">## FT358         2         1</span></span>
<span><span class="co">## FT359         2         1</span></span>
<span><span class="co">## FT360         2         1</span></span>
<span><span class="co">## FT361         1         1</span></span></code></pre>
<p>The quantified feature abundances can be accessed with the
<code>assay</code> method providing with the second argument the name of
the assay matrix to extract:</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       20171016_POOL_POS_1_105-134.mzML 20171016_POOL_POS_3_105-134.mzML</span></span>
<span><span class="co">## FT001                        3202.7445                        2285.2830</span></span>
<span><span class="co">## FT002                        3605.3915                        3183.9546</span></span>
<span><span class="co">## FT003                         744.8752                        1057.4312</span></span>
<span><span class="co">## FT004                       18126.4603                       19369.4039</span></span>
<span><span class="co">## FT005                       23243.6129                       31960.3709</span></span>
<span><span class="co">## FT006                         671.5842                         617.7545</span></span></code></pre>
<p>The <code>SummarizedExperiment</code> supports several such assay
matrices (as long as they have the same dimensions). We can thus add for
example the feature quantification excluding filled-in signal as an
additional assay matrix:</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">raw_detected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">data</span>, method <span class="op">=</span> <span class="st">"sum"</span>,</span>
<span>                                          value <span class="op">=</span> <span class="st">"into"</span>, filled <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>With that we have now two assay matrices in our result object:</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "raw"          "raw_detected"</span></span></code></pre>
<p>Analogously we could add normalized data to this object or any other
processed values. The advantage of having all the data within one object
is obvious: any subsetting of the data ensures that all assays and
annotations are subsetted correctly.</p>
<p>One final thing worth mentioning is that <em>xcms</em> result objects
keep, next to the preprocessing results, also a history of all
processing steps and all parameter objects used during the analysis. The
process history can be accessed with the <code>processHistory</code>
function.</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Overview of the performed processings</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">processHistory</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak detection </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:02 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: CentWaveParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak refinement </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:04 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: MergeNeighboringPeaksParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak grouping </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:06 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: PeakDensityParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Retention time correction </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:06 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: PeakGroupsParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak grouping </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:11 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: PeakDensityParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Missing peak filling </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:13 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: ChromPeakAreaParam </span></span>
<span><span class="co">##  MS level(s) 1</span></span></code></pre>
<p>The actual parameter object that was used to configure one particular
analysis step can be accessed with <code>processParam</code>:</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Access the parameter class for a processing step</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/ProcessHistory-class.html" class="external-link">processParam</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">processHistory</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class:  CentWaveParam </span></span>
<span><span class="co">##  Parameters:</span></span>
<span><span class="co">##  - ppm: [1] 30</span></span>
<span><span class="co">##  - peakwidth: [1]  2 10</span></span>
<span><span class="co">##  - snthresh: [1] 10</span></span>
<span><span class="co">##  - prefilter: [1]   3 100</span></span>
<span><span class="co">##  - mzCenterFun: [1] "wMean"</span></span>
<span><span class="co">##  - integrate: [1] 2</span></span>
<span><span class="co">##  - mzdiff: [1] -0.001</span></span>
<span><span class="co">##  - fitgauss: [1] FALSE</span></span>
<span><span class="co">##  - noise: [1] 0</span></span>
<span><span class="co">##  - verboseColumns: [1] FALSE</span></span>
<span><span class="co">##  - roiList: list()</span></span>
<span><span class="co">##  - firstBaselineCheck: [1] TRUE</span></span>
<span><span class="co">##  - roiScales: numeric(0)</span></span>
<span><span class="co">##  - extendLengthMSW: [1] FALSE</span></span></code></pre>
<p>All this information is also stored in the <code>metadata</code> slot
of the <code>SummarizedExperiment</code>:</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak detection </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:02 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: CentWaveParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak refinement </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:04 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: MergeNeighboringPeaksParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak grouping </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:06 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: PeakDensityParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Retention time correction </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:06 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: PeakGroupsParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak grouping </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:11 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: PeakDensityParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Missing peak filling </span></span>
<span><span class="co">##  date: Fri Sep  1 11:59:13 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2 </span></span>
<span><span class="co">##  Parameter class: ChromPeakAreaParam </span></span>
<span><span class="co">##  MS level(s) 1</span></span></code></pre>
<p>Analysis could now proceed with by e.g. normalizing the data to
remove any technical variances in the feature abundances, feature
grouping (compounding) using the <em><a href="https://bioconductor.org/packages/3.18/MsFeatures" class="external-link">MsFeatures</a></em>
package, differential abundance analysis and ultimately also annotation
and identification of the LC-MS features (e.g. with help of the <em><a href="https://bioconductor.org/packages/3.18/MetaboAnnotation" class="external-link">MetaboAnnotation</a></em>
package <span class="citation">(Rainer et al. 2022)</span>).</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="bonus-material---peak-detection-fun">Bonus material - peak detection fun<a class="anchor" aria-label="anchor" href="#bonus-material---peak-detection-fun"></a>
</h2>
<p>In this section we apply the lessons learned from previous sections,
in particular how to adapt peak detection setting on a rather noisy
<em>chromatographic</em> data. Below we load the example data from a
text file.</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"txt"</span>, <span class="st">"chromatogram.txt"</span>, package <span class="op">=</span> <span class="st">"xcmsTutorials"</span><span class="op">)</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    rt intensity</span></span>
<span><span class="co">## 1 100         0</span></span>
<span><span class="co">## 2 110         0</span></span>
<span><span class="co">## 3 120         1</span></span>
<span><span class="co">## 4 130         2</span></span>
<span><span class="co">## 5 140         4</span></span>
<span><span class="co">## 6 150         6</span></span></code></pre>
<p>Our data has two columns, one with <em>retention times</em> and one
with <em>intensities</em>. We can now create a <code>Chromatogram</code>
object from that and plot the data.</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/Chromatogram-class.html" class="external-link">Chromatogram</a></span><span class="op">(</span>rtime <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">rt</span>, intensity <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">intensity</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr</span><span class="op">)</span></span></code></pre></div>
<p><img src="xcms-preprocessing_files/figure-html/peaks-plot-1.png" width="1152"></p>
<p>There are two peaks present in the data, with the signal from the
latter being particularly noisy. The goal is now to perform the peak
detection and to identify the two peaks. A first try with the default
settings for <em>centWave</em> clearly shows that we have to tune the
parameters (note that the setting of <code>sn = 0</code> is required for
the present data set as there are not enough <em>background</em> data
points for the algorithm to estimate the noise level properly).</p>
<p>Which parameter would you now adapt to the data? What would be your
choices? Go ahead and try different settings or setting combination to
see if you can succeed in detecting the two peaks. Eventually you might
even try a different peak detection algorithm
(e.g. <code>MatchedFilterParam</code>).</p>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xchr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks</a></span><span class="op">(</span><span class="va">chr</span>, param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks-centWave.html" class="external-link">CentWaveParam</a></span><span class="op">(</span>sn <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">xchr</span><span class="op">)</span></span></code></pre></div>
<p><img src="xcms-preprocessing_files/figure-html/peaks-fail-1.png" width="1152"></p>
<p>With the default parameters centWave clearly failed to identify the
two large peaks, defining only smaller fragments of them as potential
peaks. Especially the second peak with its peculiar tri-forked shape
seems to cause troubles. This would be even for a hydrophilic liquid
interaction chromatography (HILIC), known to potentially result in noisy
odd-shaped peaks, a rather unusual peak shape. In fact, the signal we
were analyzing here is not of chromatographic origin:</p>
<p><img src="xcms-preprocessing_files/figure-html/peaks-solution-1.png" width="1152"></p>
<p>Our example data represents a panorama picture featuring mountains
from the Dolomites, the <a href="https://en.wikipedia.org/wiki/Paternkofel" class="external-link"><em>Paternkofel</em></a>
(left peak, colored red) and the famous <a href="https://en.wikipedia.org/wiki/Tre_Cime_di_Lavaredo" class="external-link"><em>Drei
Zinnen</em></a> (right tri-forked peak colored green).</p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] RColorBrewer_1.1-3          png_0.1-8                  </span></span>
<span><span class="co">##  [3] SummarizedExperiment_1.31.1 GenomicRanges_1.53.1       </span></span>
<span><span class="co">##  [5] GenomeInfoDb_1.37.3         IRanges_2.35.2             </span></span>
<span><span class="co">##  [7] MatrixGenerics_1.13.1       matrixStats_1.0.0          </span></span>
<span><span class="co">##  [9] MetaboCoreUtils_1.9.2       Spectra_1.11.9             </span></span>
<span><span class="co">## [11] MsExperiment_1.3.0          xcms_3.99.3                </span></span>
<span><span class="co">## [13] MSnbase_2.27.1              ProtGenerics_1.33.1        </span></span>
<span><span class="co">## [15] S4Vectors_0.39.1            mzR_2.35.1                 </span></span>
<span><span class="co">## [17] Rcpp_1.0.11                 Biobase_2.61.0             </span></span>
<span><span class="co">## [19] BiocGenerics_0.47.0         BiocParallel_1.35.4        </span></span>
<span><span class="co">## [21] rmarkdown_2.24              knitr_1.43                 </span></span>
<span><span class="co">## [23] BiocStyle_2.29.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] bitops_1.0-7                rlang_1.1.1                </span></span>
<span><span class="co">##  [3] magrittr_2.0.3              clue_0.3-64                </span></span>
<span><span class="co">##  [5] MassSpecWavelet_1.67.0      compiler_4.3.1             </span></span>
<span><span class="co">##  [7] systemfonts_1.0.4           vctrs_0.6.3                </span></span>
<span><span class="co">##  [9] stringr_1.5.0               pkgconfig_2.0.3            </span></span>
<span><span class="co">## [11] crayon_1.5.2                fastmap_1.1.1              </span></span>
<span><span class="co">## [13] XVector_0.41.1              utf8_1.2.3                 </span></span>
<span><span class="co">## [15] preprocessCore_1.63.1       ragg_1.2.5                 </span></span>
<span><span class="co">## [17] purrr_1.0.2                 MultiAssayExperiment_1.27.5</span></span>
<span><span class="co">## [19] xfun_0.40                   zlibbioc_1.47.0            </span></span>
<span><span class="co">## [21] cachem_1.0.8                jsonlite_1.8.7             </span></span>
<span><span class="co">## [23] progress_1.2.2              highr_0.10                 </span></span>
<span><span class="co">## [25] DelayedArray_0.27.10        prettyunits_1.1.1          </span></span>
<span><span class="co">## [27] parallel_4.3.1              cluster_2.1.4              </span></span>
<span><span class="co">## [29] R6_2.5.1                    bslib_0.5.1                </span></span>
<span><span class="co">## [31] stringi_1.7.12              limma_3.57.7               </span></span>
<span><span class="co">## [33] jquerylib_0.1.4             bookdown_0.35              </span></span>
<span><span class="co">## [35] iterators_1.0.14            igraph_1.5.1               </span></span>
<span><span class="co">## [37] splines_4.3.1               Matrix_1.6-1               </span></span>
<span><span class="co">## [39] tidyselect_1.2.0            abind_1.4-5                </span></span>
<span><span class="co">## [41] yaml_2.3.7                  doParallel_1.0.17          </span></span>
<span><span class="co">## [43] codetools_0.2-19            affy_1.79.3                </span></span>
<span><span class="co">## [45] lattice_0.21-8              tibble_3.2.1               </span></span>
<span><span class="co">## [47] plyr_1.8.8                  evaluate_0.21              </span></span>
<span><span class="co">## [49] survival_3.5-7              desc_1.4.2                 </span></span>
<span><span class="co">## [51] pillar_1.9.0                affyio_1.71.0              </span></span>
<span><span class="co">## [53] BiocManager_1.30.22         foreach_1.5.2              </span></span>
<span><span class="co">## [55] MALDIquant_1.22.1           ncdf4_1.21                 </span></span>
<span><span class="co">## [57] generics_0.1.3              rprojroot_2.0.3            </span></span>
<span><span class="co">## [59] RCurl_1.98-1.12             hms_1.1.3                  </span></span>
<span><span class="co">## [61] ggplot2_3.4.3               munsell_0.5.0              </span></span>
<span><span class="co">## [63] scales_1.2.1                glue_1.6.2                 </span></span>
<span><span class="co">## [65] lazyeval_0.2.2              MsFeatures_1.9.0           </span></span>
<span><span class="co">## [67] tools_4.3.1                 mzID_1.39.0                </span></span>
<span><span class="co">## [69] robustbase_0.99-0           QFeatures_1.11.2           </span></span>
<span><span class="co">## [71] vsn_3.69.0                  RANN_2.6.1                 </span></span>
<span><span class="co">## [73] fs_1.6.3                    XML_3.99-0.14              </span></span>
<span><span class="co">## [75] grid_4.3.1                  impute_1.75.1              </span></span>
<span><span class="co">## [77] MsCoreUtils_1.13.1          colorspace_2.1-0           </span></span>
<span><span class="co">## [79] GenomeInfoDbData_1.2.10     cli_3.6.1                  </span></span>
<span><span class="co">## [81] textshaping_0.3.6           fansi_1.0.4                </span></span>
<span><span class="co">## [83] S4Arrays_1.1.6              dplyr_1.1.2                </span></span>
<span><span class="co">## [85] AnnotationFilter_1.25.0     pcaMethods_1.93.0          </span></span>
<span><span class="co">## [87] gtable_0.3.4                DEoptimR_1.1-2             </span></span>
<span><span class="co">## [89] sass_0.4.7                  digest_0.6.33              </span></span>
<span><span class="co">## [91] SparseArray_1.1.11          multtest_2.57.0            </span></span>
<span><span class="co">## [93] memoise_2.0.1               htmltools_0.5.6            </span></span>
<span><span class="co">## [95] pkgdown_2.0.7               lifecycle_1.0.3            </span></span>
<span><span class="co">## [97] statmod_1.5.0               MASS_7.3-60</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Conley:2014ha" class="csl-entry">
Conley, Christopher J, Rob Smith, Ralf J O Torgrip, Ryan M Taylor, Ralf
Tautenhahn, and John T Prince. 2014. <span>“<span class="nocase">Massifquant: open-source Kalman filter-based XC-MS
isotope trace feature detection.</span>”</span> <em>Bioinformatics</em>
30 (18): 2636–43.
</div>
<div id="ref-gattoMSnbaseEfficientElegant2020a" class="csl-entry">
Gatto, Laurent, Sebastian Gibb, and Johannes Rainer. 2020.
<span>“<span>MSnbase</span>, <span>Efficient</span> and <span>Elegant
R</span>-<span>Based Processing</span> and <span>Visualization</span> of
<span>Raw Mass Spectrometry Data</span>.”</span> <em>Journal of Proteome
Research</em>, September. <a href="https://doi.org/10.1021/acs.jproteome.0c00313" class="external-link">https://doi.org/10.1021/acs.jproteome.0c00313</a>.
</div>
<div id="ref-Katajamaa:2006jh" class="csl-entry">
Katajamaa, Mikko, Jarkko Miettinen, and Matej Oresic. 2006. <span>“<span class="nocase">MZmine: toolbox for processing and visualization of mass
spectrometry based molecular profile data.</span>”</span>
<em>Bioinformatics</em> 22 (5): 634–36.
</div>
<div id="ref-Prince:2006jj" class="csl-entry">
Prince, John T, and Edward M Marcotte. 2006. <span>“<span class="nocase">Chromatographic alignment of ESI-LC-MS proteomics data
sets by ordered bijective interpolated warping.</span>”</span>
<em>Analytical Chemistry</em> 78 (17): 6140–52.
</div>
<div id="ref-rainer_modular_2022" class="csl-entry">
Rainer, Johannes, Andrea Vicini, Liesa Salzer, Jan Stanstrup, Josep M.
Badia, Steffen Neumann, Michael A. Stravs, et al. 2022. <span>“A
<span>Modular</span> and <span>Expandable</span> <span>Ecosystem</span>
for <span>Metabolomics</span> <span>Data</span> <span>Annotation</span>
in <span>R</span>.”</span> <em>Metabolites</em> 12 (2): 173. <a href="https://doi.org/10.3390/metabo12020173" class="external-link">https://doi.org/10.3390/metabo12020173</a>.
</div>
<div id="ref-Smith:2006ic" class="csl-entry">
Smith, Colin A, Elizabeth J Want, Grace O’Maille, Ruben Abagyan, and
Gary Siuzdak. 2006. <span>“<span class="nocase">XCMS: processing mass
spectrometry data for metabolite profiling using nonlinear peak
alignment, matching, and identification.</span>”</span> <em>Analytical
Chemistry</em> 78 (3): 779–87.
</div>
<div id="ref-Smith:2014di" class="csl-entry">
Smith, Rob, Andrew D Mathis, Dan Ventura, and John T Prince. 2014.
<span>“<span class="nocase">Proteomics, lipidomics, metabolomics: a mass
spectrometry tutorial from a computer scientist’s point of
view.</span>”</span> <em>BMC Bioinformatics</em> 15 Suppl 7 (Suppl 7):
S9.
</div>
<div id="ref-Tautenhahn:2008fx" class="csl-entry">
Tautenhahn, Ralf, Christoph Böttcher, and Steffen Neumann. 2008.
<span>“<span class="nocase">Highly sensitive feature detection for high
resolution LC/MS.</span>”</span> <em>BMC Bioinformatics</em> 9 (1): 504.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
